packages/api/src/rich-text/rich-text.ts

```diff
@@ -0,0 +1,359 @@
+        ent.index.end = removeStartIndex
+      }
+      // scenario D (entirely inner)
+      else if (
+        removeStartIndex >= ent.index.start &&
+        removeEndIndex <= ent.index.end
+      ) {
+        // move end by num removed
+        ent.index.end -= numCharsRemoved
+      }
+      // scenario E (partially before)
+      else if (
+        removeStartIndex < ent.index.start &&
+        removeEndIndex >= ent.index.start &&
+        removeEndIndex <= ent.index.end
+      ) {
+        // move start to remove-start index, move end by num removed
+        ent.index.start = removeStartIndex
+        ent.index.end -= numCharsRemoved
+      }
+      // scenario F (entirely before)
+      else if (removeEndIndex < ent.index.start) {
+        // move both by num removed
+        ent.index.start -= numCharsRemoved
+        ent.index.end -= numCharsRemoved
+      }
+    }
+
+    // filter out any facets that were made irrelevant
+    this.facets = this.facets.filter((ent) => ent.index.start < ent.index.end)
+    return this
+  }
+
+  /**
+   * Detects facets such as links and mentions
+   * Note: Overwrites the existing facets with auto-detected facets
+   */
+  async detectFacets(agent: AtpAgent) {
+    this.facets = detectFacets(this.unicodeText)
+    if (this.facets) {
+      for (const facet of this.facets) {
+        if (AppBskyRichtextFacet.isMention(facet.value)) {
+          const did = await agent
+            .resolveHandle({ handle: facet.value.did })
+            .catch((_) => undefined)
+            .then((res) => res?.data.did)
+          facet.value.did = did || ''
+        }
+      }
+    }
+  }
+
+  /**
+   * Detects facets such as links and mentions but does not resolve them
+   * Will produce invalid facets! For instance, mentions will not have their DIDs set.
+   * Note: Overwrites the existing facets with auto-detected facets
+   */
+  detectFacetsWithoutResolution() {
+    this.facets = detectFacets(this.unicodeText)
+  }
+}
+
+function entitiesToFacets(text: UnicodeString, entities: Entity[]): Facet[] {
+  const facets: Facet[] = []
+  for (const ent of entities) {
+    if (ent.type === 'link') {
+      facets.push({
+        $type: 'app.bsky.richtext.facet',
+        index: {
+          start: text.utf16IndexToUtf8Index(ent.index.start),
+          end: text.utf16IndexToUtf8Index(ent.index.end),
+        },
+        value: { $type: 'app.bsky.richtext.facet#link', uri: ent.value },
+      })
+    } else if (ent.type === 'mention') {
+      facets.push({
+        $type: 'app.bsky.richtext.facet',
+        index: {
+          start: text.utf16IndexToUtf8Index(ent.index.start),
+          end: text.utf16IndexToUtf8Index(ent.index.end),
+        },
+        value: { $type: 'app.bsky.richtext.facet#mention', did: ent.value },
+      })
+    }
+  }
+  return facets
+}
+
+function cloneDeep<T>(v: T): T {
+  if (typeof v === 'undefined') {
+    return v
+  }
+  return JSON.parse(JSON.stringify(v))
```

Comment:
Not sure if it fits the RN environment but there's also `structuredClone()` for this, which is pretty friendly.